crdutil: Add feature to also apply also single files

tobiasgiese opened the pull request.

tobiasgiese commented on pkg/crdutil/crdutil.go:
```diff
@@ -50,21 +50,29 @@ func (s *StringList) Set(value string) error {
 
 func initFlags() {
 	flag.Var(&crdsDir, "crds-dir", "Path to the directory containing the CRD manifests")
+	flag.Var(&crds, "crds", "List of single CRDs files with CRD manifests to apply")
```
Was not sure if `--crds` or `--crds-file`

tobiasgiese replied:
Changed to `--crds-file`

shivamerla commented on pkg/crdutil/crdutil.go:
```diff
@@ -49,20 +49,28 @@ func (s *StringList) Set(value string) error {
 
 func initFlags() {
-	flag.Var(&crdsDir, "crds-dir", "Path to the directory containing the CRD manifests")
+	flag.Var(&crdsDirs, "crds-dir", "Path to the directory containing the CRD manifests")
```
given that this is crd specific util, we can just have one flag `-f` `--filename` which can be a file or directory. We can also use `-r` `--recursive` option to decide on recursively processing all yaml files in the directory which by default is false. This aligns with kubectl command as well.

shivamerla commented on pkg/crdutil/crdutil.go:
```diff
@@ -87,11 +95,15 @@ func EnsureCRDsCmd() {
 // walkCrdsDir walks the CRDs directory and applies each YAML file.
 func walkCrdsDir(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface) error {
-	for _, crdDir := range crdsDir {
+	for _, crdDir := range crdsDirs {
```
consider files with both `.yaml` or .yml` suffix

shivamerla commented on pkg/crdutil/crdutil.go:
```diff
@@ -49,20 +49,28 @@ func (s *StringList) Set(value string) error {
+
+	for _, crd := range crdsFiles {
+		if _, err := os.Stat(crd); os.IsNotExist(err) {
+			log.Fatalf("CRD file %s does not exist", crd)
```
`ApplyCRDs()` seems more appropriate for the below function

shivamerla commented on pkg/crdutil/crdutil.go:
```diff
@@ -108,7 +120,17 @@ func walkCrdsDir(ctx context.Context, crdClient v1.CustomResourceDefinitionInter
+	return nil
+}
+
+func applyCrdFiles(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface) error {
```
For the `applyCRD` call, err can be intermediate (i.e other than NOT_FOUND when the CRD is already created) and curCRD can be nil. need to handle that case.

tobiasgiese replied:
Good catch 👍🏻

tobiasgiese commented:
Thanks for the extensive review @shivamerla, I should have handled all findings in [this diff](https://github.com/NVIDIA/k8s-operator-libs/compare/7b4a23b8e084468a9a29e7815946e237064c9da2..43b027cb970d799451dff977291685d037065e9b).

tobiasgiese commented:
Tanks for the extensive review @shivamerla. I should have implemented all findings [in this diff](https://github.com/NVIDIA/k8s-operator-libs/compare/7b4a23b8e084468a9a29e7815946e237064c9da2..776bbdbd1088985271b40af465d6b7315798c536).

tobiasgiese commented on pkg/crdutil/crdutil.go:
```diff
@@ -14,6 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+//nolint:depguard,lll
```
```
pkg/crdutil/crdutil.go:62: the line is 148 characters long, which exceeds the maximum of 120 characters. (lll)
		"Process the directory used in -f, --filename recursively. Useful when you want to manage related manifests organized within the same directory.")
pkg/crdutil/crdutil.go:29:2: import 'github.com/spf13/pflag' is not allowed from list 'main' (depguard)
	"github.com/spf13/pflag"
	^
make: *** [lint] Error 1
```

IMO both is fine to ignore

tobiasgiese replied:
But if we say that we don't want to have the direct dependency to pflag I can remove it and use stdlib flag

vasrem replied:
`lll` can be removed by splitting the string in multiple lines. I think that's easy to do.

For the `depguard`, maybe worth adding it here instead https://github.com/NVIDIA/k8s-operator-libs/blob/main/.golangci.yaml#L104.

In any case, if we can go with flag (like we used to), would be better to avoid yet another dependency.

tobiasgiese replied:
> lll can be removed by splitting the string in multiple lines. I think that's easy to do.

Maybe I can also remove the second sentence. It's just a copy from `kubectl`, but we can adjust it IMO.

> For the depguard, maybe worth adding it here instead https://github.com/NVIDIA/k8s-operator-libs/blob/main/.golangci.yaml#L104.

Yes, also thought about this, but we should rather discuss why we need this anyway?
All dependencies, if they are indirect or direct will be indirect dependencies in the consumers go.mod.
IMO we don't need that `depguard` check.

Repro with the following:

```
❯ ll
total 28K
-rw-r--r-- 1 tgiese wheel 2.8K Nov 27 14:19 go.mod
-rw-r--r-- 1 tgiese wheel  18K Nov 27 14:19 go.sum
-rw-r--r-- 1 tgiese wheel  119 Nov 27 14:16 main.go
❯ cat main.go
package main

import (
	"github.com/NVIDIA/k8s-operator-libs/pkg/crdutil"
)

func main() {
	crdutil.EnsureCRDsCmd()
}
```

https://github.com/NVIDIA/k8s-operator-libs/blob/92553f3478d75f72c6ad58a8b8b329866bed5275/go.mod#L9
require github.com/spf13/pflag v1.0.5

```
❯ go mod tidy; grep -e replace -e pflag go.mod
	github.com/spf13/pflag v1.0.5 // indirect
replace github.com/NVIDIA/k8s-operator-libs => github.com/tobiasgiese/k8s-operator-libs v0.0.0-20241127003902-92553f3478d7
```

and without the replace (i.e., with pflag indirect dep in the lib)
```
❯ go mod tidy; grep -e replace -e pflag go.mod
	github.com/spf13/pflag v1.0.5 // indirect
❯
```

vasrem replied:
> All dependencies, if they are indirect or direct will be indirect dependencies in the consumers go.mod.
IMO we don't need that depguard check.

In general, it's good to avoid importing deps on packages that are imported by other consumers to avoid diamond dependencies. In that particular case, you proved that the `github.com/NVIDIA/k8s-operator-libs/pkg/crdutil` is importing some package that already imports pflag, so in that case, it's not that bad to import pflag. But this is not the case for any package I believe.

tobiasgiese replied:
Added `github.com/spf13/pflag` to the depguard allow list and cut the flag description

vasrem commented on pkg/crdutil/crdutil.go:
```diff
@@ -171,6 +188,9 @@ func applyCRD(
 			return fmt.Errorf("create CRD %s: %w", crd.Name, err)
 		}
 	} else {
+		if err != nil {
```
nit: can we return in L189 and remove the else block here?

vasrem commented on pkg/crdutil/crdutil.go:
```diff
@@ -84,38 +92,47 @@ func EnsureCRDsCmd() {
+				return nil
+			}
+			// If not recursive we want to only apply the CRDs in the top-level directory.
+			if !recursive && parentDir.IsDir() && filepath.Dir(path) != strings.TrimRight(crdDir, "/") {
```
should we add a test case for this new branch or should these all be included in the "TODO add unit tests" comment you have added?

tobiasgiese replied:
Let me take another look at the tests, I think I can easily test the `walkCRDs` func

vasrem reviewed (commented):
LGTM overall, just one comment on the linter stuff.

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -84,38 +92,47 @@ func EnsureCRDsCmd() {
 				return err
 			}
-			if info.IsDir() || filepath.Ext(path) != ".yaml" {
+			if info.IsDir() {
```
Question: if `path` is a directory should we not decend into it if recursive is set?

tobiasgiese replied:
But we walk through the files in the directory already and we are only interested in files

tobiasgiese replied:
See this test https://github.com/NVIDIA/k8s-operator-libs/pull/58#issuecomment-2502284586

elezar replied:
I think that's fine. Thanks for clarifying.

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -49,26 +51,32 @@ func (s *StringList) Set(value string) error {
 var (
-	crdsDir StringList
+	files     []string
+	recursive bool
```
Does it make sense to separate the handling of multiple non-recursive paths (including files) from the addition of recursive handling?

tobiasgiese replied:
AFAIK `kubectl` is also not separating it, so it is totally fine to keep it pretty open IMO. But also fine to me to split it somehow (you mean like: if multiple files are passed the recursive flag is not possible?)

shivamerla replied:
If the network-operator is not released with this change, better to just update as it will be first use of this package.

elezar replied:
I actually just meant as separate PRs. From https://github.com/NVIDIA/gpu-operator/pull/1137 it seems that we don't use recursive mode. Is it used in the network operator?

tobiasgiese replied:
No, it is not used by the network-operator. Maybe it makes sense to remove this new feature from the PR, you are right. The request adding it to this PR came through https://github.com/NVIDIA/k8s-operator-libs/pull/58#discussion_r1859239149 (at least that's how I understood it)

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -49,26 +51,32 @@ func (s *StringList) Set(value string) error {
-		log.Fatalf("CRDs directory is required")
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+	pflag.StringSliceVarP(&files, "filename", "f", files,
+		"The files that contain the configurations to apply.")
```
This usage string does not seem to accurately represent the intent of the argument(s). From the PR description, I assume that each `filename` represents a path. Each of these can be a directory or a YAML file path. For directories, the YAML files in the directory are processed.

tobiasgiese replied:
The flag names, the logic and also the description is copied from `kubectl apply` :)

tobiasgiese replied:
But to explain this a bit more:

- the old behavior was indeed recursive by default, but without the possibility to decide this
- the new behavior only handles yaml files in the same directory, if a directory is passed
- with setting the recursive flag all directories will be handled recursively

Let me add tests for this to make this pattern clear

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -49,26 +51,32 @@ func (s *StringList) Set(value string) error {
+	pflag.StringSliceVarP(&files, "filename", "f", files,
+		"The files that contain the configurations to apply.")
+	pflag.BoolVarP(&recursive, "recursive", "R", false,
+		"Process the directory used in -f, --filename recursively. Useful when you want to manage related manifests organized within the same directory.")
```
Is this not already handled without the `-R` flag? (This was the previous behaviour).

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -49,26 +51,32 @@ func (s *StringList) Set(value string) error {
 )
 
 func initFlags() {
-	flag.Var(&crdsDir, "crds-dir", "Path to the directory containing the CRD manifests")
```
Are we concerned about this as a breaking change since the command line arguments are changed?

tobiasgiese replied:
I was also thinking about this but this pkg is pretty new and we are only using it for the network-operator currently

shivamerla replied:
If the network-operator is not released with this change, better to just update as it will be first use of this package.

tobiasgiese replied:
The network-operator 24.10.0 has been released 2 days ago with the change in it, see https://github.com/Mellanox/network-operator/blob/v24.10.0/deployment/network-operator/templates/upgrade-crd-hook.yaml#L84.

But IMO it is totally fine to change it, because it is a Helm pre-install hook. If it fails, it fails the installation/upgrade and the user has to use the latest Helm chart. There is no dependency to the runtime which can cause the operator to fail.

WDYT @e0ne @rollandf

elezar replied:
If there is a concern, we could support both -- marking `crds-dir` as deprecated -- for a release. Happy with the change since this is effectively an internal API.

tobiasgiese replied:
We can also revert it to `--crds-dir`. Both `crds-dir` and `filename` does not contain both files and dirs, so it is not 100% explicit anyway :)
Request came through https://github.com/NVIDIA/k8s-operator-libs/pull/58#discussion_r1859239149

tobiasgiese replied:
@shivamerla WDYT? Should we remove the flag rename and recursive flag (see https://github.com/NVIDIA/k8s-operator-libs/pull/58#discussion_r1867228854) out of this PR and change it in a follow-up? I'd say this will reduce the size of this PR.

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -84,38 +92,47 @@ func EnsureCRDsCmd() {
-	for _, crdDir := range crdsDir {
+// walkCRDs walks the CRDs directory and applies each YAML file.
+func walkCRDs(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface) error {
+	for _, crdDir := range files {
```
I understand that `files` is global, but does it makes sense to pass it in as an argument instead?

tobiasgiese replied:
Same for the `recursive` bool

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -84,38 +92,47 @@ func EnsureCRDsCmd() {
-func walkCrdsDir(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface) error {
-	for _, crdDir := range crdsDir {
+// walkCRDs walks the CRDs directory and applies each YAML file.
+func walkCRDs(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface) error {
```
As a question: Does it makes sense to separate filename collection from applying the CRDs? That is to say that we would have logic that accepts the flags and returns a slice of YAML file paths. The logic to apply the CRDs would then accept a slice to YAML file paths. This should make things simpler to test and also allow us to properly de-duplicate if a user specifies a file multiple times, for example.

shivamerla reviewed (approved):
LGTM! Thanks

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -37,38 +38,31 @@ import (
+	for _, crdDir := range filenames {
 		if _, err := os.Stat(crdDir); os.IsNotExist(err) {
-			log.Fatalf("CRDs directory %s does not exist", crdsDir)
+			log.Fatalf("CRDs directory %s does not exist", filenames)
```
should this be:
```suggestion
			log.Fatalf("path does not exist: %s", cdrDir)
```

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -37,38 +38,31 @@ import (
 	}
 
-	for _, crdDir := range crdsDir {
+	for _, crdDir := range filenames {
```
```suggestion
	for _, path := range filenames {
```

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -84,38 +78,59 @@ func EnsureCRDsCmd() {
 
-	if err := walkCrdsDir(ctx, client.ApiextensionsV1().CustomResourceDefinitions()); err != nil {
-		log.Fatalf("Failed to apply CRDs: %v", err)
+	dirsToApply, err := walkCRDs(recursive, filenames)
```
```suggestion
	pathsToApply, err := collectYamlPaths(recursive, filenames)
```

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -84,38 +78,59 @@ func EnsureCRDsCmd() {
-// walkCrdsDir walks the CRDs directory and applies each YAML file.
-func walkCrdsDir(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface) error {
-	for _, crdDir := range crdsDir {
+// walkCRDs walks the CRDs directory and applies each YAML file.
+// TODO: add unit test  for this function.
+func walkCRDs(recursive bool, crdDirs []string) ([]string, error) {
```
```suggestion
// collectYAMLPaths processes a list of paths and returns all YAML files.
// If `recursive` is specified directories in the list are searched recursively for YAML files.
// TODO: add unit test  for this function.
func collectYAMLPaths(recursive bool, paths []string) ([]string, error) {
```

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -84,38 +78,59 @@ func EnsureCRDsCmd() {
-// applyCRDsFromFile reads a YAML file, splits it into documents, and applies each CRD to the cluster.
-func applyCRDsFromFile(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface, filePath string) error {
+// applyCRDs reads a YAML file, splits it into documents, and applies each CRD to the cluster.
+func applyCRDs(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface, filePath string) error {
```
This rename wasn't strictly required, correct?

tobiasgiese replied:
No, it's not required. It was a review finding https://github.com/NVIDIA/k8s-operator-libs/pull/58#discussion_r1859242624

elezar commented on pkg/crdutil/crdutil.go:
```diff
@@ -170,14 +185,19 @@ func applyCRD(
 		if err != nil {
 			return fmt.Errorf("create CRD %s: %w", crd.Name, err)
```
Note: we already add `crd.Name` at the caller. Could we just return:
```suggestion
			return fmt.Errorf("create CRD: %w", err)
```
(and similarly for `Update` below)?

tobiasgiese commented:
just as a note: will work on it next week again

tobiasgiese commented:
@elezar @shivamerla I implemented all findings and also removed the `--force` flag and reverted the flag rename from `--filename` back to `--crds-dir`. We can think of deprecating the flag and use `--filename` in a follow-up PR. Additionally, the force feature can/will be also implemented in a follow-up PR - if needed.

If you both accept the changes I will squash both commits.
